Show diffs on multiple changed goals; match old and new goal info

jfehrle opened the pull request:
**Kind:** bug fix / feature.

Currently the diff code only compares the first current goal of the old and new proof states.  While this is correct often enough to be useful, it doesn't handle the cases where:
- Multiple goals are changed or created with one tactic, such as "split."
- The order of the goals is changed, e.g. with "all: swap 1 3."

Note: the first commit changes the behavior of the recently-added Stm.get_prev_proof function.  This is so that tactics and commands that don't change the underlying proof, such as "all : swap 1 3", still show the same diffs from the last proof-changing tactic, though in a different spot in the output.

I hope we can get this into 8.9.

For example:

![image](https://user-images.githubusercontent.com/1253341/44053218-d35bd59c-9ef3-11e8-95b1-2449df37aa21.png)

jfehrle commented:
@ejgallego @Zimmi48 This solves the underlying problem of #7653.

jfehrle commented on printing/proof_diffs.ml:
```diff
@@ -317,31 +333,244 @@ let hyp_list_to_pp hyps =
+  let pprf = Proof.partial_proof p in
+  (* todo: When does pprf have multiple elements?  Can it have zero elements? *)
+  let t = List.hd pprf in
+  let sigma, env = Pfedit.get_current_context ~p () in
```
I'm pretty sure zero elements is not possible here.  If it's possible to have multiple elements, I'd like to know how to generate such a case so I can try it.

Zimmi48 replied:
@ejgallego can probably answer this. I think there is support for multiple proofs that is only used by the Derive plugin AFAIR.

jfehrle replied:
Thanks.  BTW, I tried `Set Nested Proofs Allowed.' and entering a new Goal but I still didn't get multiple entries in pprf.

SkySkimmer replied:
It might be for
~~~coq
Lemma foo : A with bar : B.
Proof. exact A_val. exact B_val. Qed.
~~~
I didn't check so this is just a conjecture.

jfehrle replied:
Ah, can you send me the exact text to try?

![image](https://user-images.githubusercontent.com/1253341/44080165-12636428-9f60-11e8-8be6-a23c87a2f6c6.png)

Zimmi48 replied:
``` coq
Variables A B : Set.
Inductive tree : Set := node : A -> forest -> tree

with forest : Set :=
| leaf : B -> forest
| cons : tree -> forest -> forest.

Theorem tree_size : tree -> nat with forest_size : forest -> nat.
```

jfehrle replied:
Thanks for the example text.  I still find pprf has only one element.  Perhaps it is a vestige and the code no longer generates pprf with more than one element.

herbelin replied:
It seems that the ability to have several goals in `partial_proof` is not used in general. It is used in the `derive` plugin but I did not investigate more.

jfehrle commented:
The ci-fiat-crypto failure looks unrelated.

@herbelin, most of the changes are in the printing directory, so perhaps you should be the first reviewer?  Thanks.

jfehrle commented:
I would have preferred a different approach myself.  Ideally, associating goals in the new proof state with their antecedents would be handled in the lower level code when a tactic is applied.  I was advised that this was not a simple change.  My understanding is that tactics provide the new proof state as a whole, so the associations are not trivially available even at that level, so a good bit of refactoring would be needed.  I thought it was better to implement this outside the lower level code so it's much less likely to raise any critical issue (e.g. the diff code is in try blocks and the user has to enable the feature), particularly as someone relatively new to Coq and who has not worked at all with the lower level code.

jfehrle commented on printing/printer.mli:
```diff
@@ -171,23 +171,25 @@ val pr_transparent_state   : transparent_state -> Pp.t
+val pr_goal                : ?diffs:bool -> ?prev_g_s:(goal sigma) -> goal sigma -> Pp.t
 
-(** [pr_subgoals ~pr_first ~prev_proof pp sigma seeds shelf focus_stack unfocused goals]
+module GoalMap = Evar.Map
```
Got it.

What I really wanted was to define a Map module in goal.mli.  But that opens Util, which defines its own conflicting Map module--it took me a couple hours to realize that.  I wasn't keen on moving my Map definition before the `open`.

I also experimented for a couple hours without success with variations such as:

```
module Map : sig include Map.S with type key = goal end      (* in .ml file *)
module Map = Map.Make(struct type t = goal let compare = Evar.compare end)      (* in .mli file *)
```

It's not easy to figure out what might work from the OCaml doc or the other sources I checked on the internet.

jfehrle commented on stm/stm.ml:
```diff
@@ -1231,9 +1231,23 @@ end = struct (* {{{ *)
+      | Some cp ->
+        let did = ref id in
+        let rv = ref np in
+        while
```
Most imperative languages have a `do ... while (condition)` that would work nicely here that OCaml doesn't support.  What I did is an analog.  I can add a ref bool `done` flag and move the test into the body.

But if you think there shouldn't be ref variables as well, what would you suggest replacing them with?

jfehrle commented:
Updated.  Let me know if you have more comments.  Thanks.

jfehrle commented:
Fixed a minor bug in the diff printing adn did a little refactoring for clarity in the code I've added.

herbelin commented on printing/printer.ml:
```diff
@@ -494,17 +494,17 @@ let pr_transparent_state (ids, csts) =
 (* display complete goal
- prev_gs has info on the previous proof step for diffs
- gs has info on the current proof step
+ prev_g_s has goal+sigma on the previous proof step for diffs
```
Is this now `og_s`?

herbelin commented on printing/proof_diffs.ml:
```diff
@@ -14,8 +14,8 @@ Proof General requires minor changes to make the diffs visible, but this code
 
-Diffs are computed for the hypotheses and conclusion of the first goal between
-the old and new proofs.
+Diffs are computed for the hypotheses and conclusion of the each goal in the new
```
the each -> each?

herbelin commented on printing/proof_diffs.ml:
```diff
@@ -317,31 +327,277 @@ let hyp_list_to_pp hyps =
     str "============================" ++ cut () ++
     concl_pp);;
+
+(* CODE TO DETERMINE WHICH GOALS TO COMPARE BETWEEN THE OLD AND NEW PROOFs *)
```
Why uppercase?

herbelin commented on printing/proof_diffs.ml:
```diff
@@ -317,31 +327,277 @@ let hyp_list_to_pp hyps =
+    invalid_arg "Array.iter2: arrays must have the same length"
+  else
+    for i = 0 to Array.length a - 1 do f (Array.unsafe_get a i) (Array.unsafe_get b i) done
+
```
This is already in clib/cArray.ml (available as `Array.iter2`).

jfehrle commented:
I've made changes addressing your comments.  Thanks for the review!

@gares, any comments on the stm change?

gares commented on stm/stm.ml:
```diff
@@ -1231,9 +1231,22 @@ end = struct (* {{{ *)
+      | Some cp ->
+        let did = ref id in
+        let rv = ref np in
+        let _done = ref false in
```
why `_done`? The variable is used, did you get a warning?

jfehrle replied:
You mean why the name `_done` and not `done`?  `done` is a keyword, it can't be used for a variable name, right?

Zimmi48 replied:
Indeed, but then find another name. @gares is right: pre-pending `_` is conventionally reserved to unused variables.

herbelin replied:
Note: for a function called `match`, we used `match_`!

jfehrle replied:
Ah, I was unaware of the convention.

ejgallego commented on printing/proof_diffs.ml:
```diff
@@ -317,31 +324,270 @@ let hyp_list_to_pp hyps =
+open Names
+open CAst
+
+let term_to_name x =
```
I am not sure `glob_expr` should be used here; this PR certainly deserves a deeper look.

jfehrle replied:
This code was based on reverse engineering `show_proof`, the code for the `Show Proof` command, which is based entirely on same type that `x` has.  If that's good for `Show Proof`, why wouldn't it be good here?

BTW `x` is of type `'a glob_constr_r`; `glob_expr` is not (no longer?) a defined type in source tree.

ejgallego replied:
> If that's good for Show Proof, why wouldn't it be good here?

I'm much afraid the code for `Show Proof` may not be good.

jfehrle replied:
What data structure would you suggest using instead?

This code is far simpler than `Show Proof`--it's just a somewhat verbose tree traversal constructed mechanically from the type definition.  If `Show Proof` is using this data structure, it seems overzealous to say that diffs shouldn't.  If and when `Show Proof` is restructured, I can make the corresponding changes here.

The consequences of a failure to match or an incorrect match here are limited--the diff code will catch the failure to match, then print an error message and the proof step without diffs.  If there's an incorrect match, the diffs will look a bit odd, but the user can turn off diffs if that's distracting--and I'll make improvements.  `match_goals` is used only when multiple old goals have been rewritten, I suspect most proof steps don't do that.

ejgallego replied:
I didn't see where show proof uses a glob_constr , care to point that out?

ejgallego replied:
I stilll don't see where the `glob_constr` is used.

jfehrle replied:
> I stilll don't see where the `glob_constr` is used.

Sorry, I didn't see this question until just now.  In Constrextern.extern, the `c` in `match c with` is of type `glob_constr_r`.

jfehrle commented:
I think "heuristics" is the wrong term; if my assumptions are correct, this is a full solution.

I added a description of the matching in comments just before Proof_diffs.make_goal_map_i (and a reference to them in the commit message).  I also adjusted some variable names to be more precise about things that are goal ids versus evar names.

Also, `_done` is now `done_.

jfehrle commented:
@herbelin Would you mark your review as "Approved"?

Still waiting to hear back from @ejgallego.  There's not a lot of time left before 8.9 beta if significant changes are needed.

jfehrle commented on proofs/proof.ml:
```diff
@@ -452,3 +452,12 @@ module V82 = struct
+    List.fold_left (fun s g -> Goal.Set.add g s) set gs in
+  let (goals,stack,shelf,given_up,_) = proof p in
+    let set = add goals Goal.Set.empty in
+    let set = List.fold_left (fun s gs -> let (g1, g2) = gs in add g1 (add g2 set)) set stack in
```
This is a set, not a map.  You mean I should update lib/cSet.ml?  To simplify line 461 only?  `stack` is a list of (list of goals, list of goals)--that seems very specific to this routine, hard to imagine that would be reused elsewhere if it was in cSet.

jfehrle commented:
Pushed an updated description.

herbelin commented:
@ejgallego: you seem to be the one who looked the most closely at this PR, do you want to be the assignee?

herbelin commented:
Maybe should I be the assignee...

If I understand correctly, all local comments have been addressed but the remark of @ppedrot about the API level at which the feature should be implemented is still open, is that right?

If yes, could @ppedrot explain what he has in mind so as to understand if a modified implementation is worth in the longer term?

As of today, the current implementation is anyway the only implementation available, so why not experimenting with it as it is?

If this position is acceptable for everyone, I shall merge by Monday (@silene: this shall be ok?).

ppedrot commented:
My remark does not prevent merging as-is. I don't have a better proposal currently, but eventually this code should be moved upwards to the UI where it really belongs.

ejgallego commented:
> If I understand correctly, all local comments have been addressed but the remark of @ppedrot about the API level at which the feature should be implemented is still open, is that right?

@herbelin So far most of the code changes in this PR do seem like an improvement, however, the use of `glob_expr` in `Proof_diffs` seems strange and I would like to understand it better before we merge. Unfortunately I am a bit short on time to be able to look at that code in detail.

That being said, I don't expect that issue to be a blocker for 8.9, so if you think the code there seems OK feel free to merge.

ejgallego commented:
> eventually this code should be moved upwards to the UI where it really belongs.

What code? IMHO the diff code does belong in the printing subsystem as editors do have a hard time making proper diffs without having access to the underlying AST.

herbelin commented:
> That being said, I don't expect that issue to be a blocker for 8.9, so if you think the code there seems OK feel free to merge.

@ejgallego: You probably looked at the code more than I did. So, the question is more to me whether the PR improves the diff feature or not even if the implementation can be made differently. For the first question, is the answer correctly a yes? Then, wrt the implementation, will it be possible that @ppedrot, you and @jfehrle find a protocol to make the code evolve? Eventually, I can spend some time to understand the issues which @ppedrot and you are raising but I'm sorry that my experience in UI is really too weak to have a proper opinion myself yet.

ejgallego commented:
> You probably looked at the code more than I did. So, the question is more to me whether the PR improves the diff feature or not even if the implementation can be made differently.

I am concerned about some of the code, in the diffs file; the rest seems fine. My concern is big enough as to make this wait unless some other person can have a look at the use of `Glob_term`.

I feel like @ppedrot has some other concerns.

jfehrle commented:
Are you discussing who feels comfortable doing the merge or whether this can be merged as-is?

If neither of you (@herbelin, @ejgallego) feel comfortable merging this, then perhaps @ppedrot would be willing to do so since he knows the lower level code better.


Regarding whether the code can be merged as-is:

> I feel like @ppedrot has some other concerns.

He approved the PR 3 weeks ago, so his concerns were not that great.  And more recently he said "My remark does not prevent merging as-is."

@ejgallego: Given:
* your earlier statement "I am fine with the PR, it seems to improve code in a key few places"
* that this change is low risk (the feature can be turned off),
* that there are no specific comments I can respond to and 
* that I've offered to rework it in the future if necessary,

why delay this fix/improvement any longer?  Let's not make the best the enemy of the good (as Voltaire put it).

ejgallego commented:
@jfehrle I am sorry if I haven't been clear, but my request about why the code is using glob_constr has still not been answered and that's a critical concern.

The code you pointed out above has nothing to do with the use of an internal AST which worries me.

jfehrle commented:
I have no idea why that would be a critical concern.  Would you give at least a little detail so I understand your concern?  What are the alternatives?  Would it help to move some of these routines to a different file?

> The code you pointed out above has nothing to do with the use of an internal AST which worries me.

What does "The code you pointed out above" refer to?

ejgallego commented:
See this comment:

https://github.com/coq/coq/pull/8247#discussion_r211646080

jfehrle commented:
> I am not sure `glob_expr` should be used here; this PR certainly deserves a deeper look.

What would address your concerns?  Further review by @ppedrot or someone else on this point?  I don't know what you're thinking here so I don't know what to do about it.

silene commented:
> If this position is acceptable for everyone, I shall merge by Monday (@silene: this shall be ok?).

The date does not matter much, since v8.9 was branched last week. This does not seem like a fix, so I will backport it only if developers make a good case for it.

Zimmi48 commented:
@silene 
>This does not seem like a fix, so I will backport it only if developers make a good case for it.

This is a refinement of a new (imperfect) feature that has never been part of any release so it seems like a good candidate for backporting.

ejgallego commented:
> What would address your concerns?

My question is, "why are you using a `glob_expr` instead of a `constr_expr`". Why do you need that printing function there.

You already tried to give an answer by pointing to some code that does nothing like yours, so that's not a very good one I'd dare to say.

ppedrot commented:
Just to be clear: this code is wrong, but AFAIU there is no way to make it correct, because it relies on false premises. So diffing subgoals is just a horrible heuristic, which is OK because this is just UI. That's the reason why I'd like to see this code in another component I don't want to have to care about.

ejgallego commented:
> Just to be clear: this code is wrong, but AFAIU there is no way to make it correct, because it relies on false premises. So diffing subgoals is just a horrible heuristic

I think you are a bit radical here. Diffing in general is a heuristic anyways; so IMO the real question here is: "what is the heuristic? Is it implemented in a reasonable way?"

Note that we could move this to some upper layer, but printing seems OK. `toplevel` uses it so it cannot go too high anyways.

jfehrle commented:
> My question is, "why are you using a `glob_expr` instead of a `constr_expr`".

I'm completely new to this part of the code and unfamiliar with the difference between the two.  I don't know enough about them to respond intelligently.  I was able to reverse-engineer something based on glob_constr (not glob_expr).  I can look at changing it if it's important, but I probably would need some help getting pointed in the right direction.

> Why do you need that printing function there.

Are you referring to show_proof?  That's just what I used to figure out how the data structure works.

ejgallego commented:
Do you mean that the code is not actually used?

jfehrle commented:
> Do you mean that the code is not actually used?

The diffs code does not call show_proof, if that is your question.

ejgallego commented:
Oh indeed I see. Note that coq is a mature system thus new dead debug code is not allowed in the production release.

Additionally it had the effect of delaying this PR.

ejgallego commented:
Sorry Jim, I am totally lost. I guess I'll have to wait until I have time for a proper review.

jfehrle commented:
Any idea when you'll have time for a proper review?
I'd be happy to do a Skype call at that time; that might help get us into sync on the PR.

ejgallego commented:
> Any idea when you'll have time for a proper review?

Early next week for sure. I think beta is still a couple of weeks away so we should be good.

ejgallego commented:
Anyways let me state my position here, if I had to chose whether I recommend this going into 8.9 or waiting for 8.10 in this current state I'd recommend getting the PR in, but I think it'd be wise to do one more pass before merging, in particular `Glob_constr` shouldn't be used in `Proof_diffs`.

jfehrle commented:
> It has never been part of any release, but it does modify some external interfaces. For example, this pull request adds an og_s argument to various exported functions. I have absolutely no idea what og_s means and it is not documented anywhere, as far as I can tell.

Certainly I can add a description of the og_s arguments.  I would observe, though, that
* Only one out of about 50 functions defined in `printer.mli` has any description of the purpose or arguments.  (Though it's not unreasonable to require more documentation for newly-added items.)
* I expect there won't be a lot of people writing code against these changed external interfaces.

Zimmi48 commented:
>Though it's not unreasonable to require more documentation for newly-added items

:+1: Fortunately, the PR review system allows us to enforce better quality on new code.

jfehrle commented:
The example in issue https://github.com/coq/coq/issues/8481 breaks the tree-matching heuristic.  It doesn't handle the `1,4: pose True.` case because the new proof is not just the old proof with substitutions for one or more goals.  (If the `pose` is done in 2 steps, the match succeeds with a simpler heuristic: if only one goal is added and one goal is removed between the steps, then the new goal is a modification of the old goal.)

```
Unnamed_thm < repeat apply conj.
  :
Unnamed_thm < Show Proof.
(conj ?Goal (conj ?Goal0 (conj ?Goal1 ?Goal2)))

Unnamed_thm < 1,4:pose True.
Warning:
Diff failure: Unable to match goals between old and new proof states (5)
Showing results without diff highlighting
  :
Unnamed_thm < Show Proof.
(conj (let P := True in ?Goal1)
   (conj ?Goal (conj ?Goal0 (let P := True in ?Goal2))))
```

jfehrle commented:
> The example in issue #8481 breaks the tree-matching heuristic.

I spoke too soon.  It was an easily-fixable bug.

jfehrle commented:
I pushed the latest, updating comments in printer.mli as requested and fixing the bug I mentioned.

@ejgallego, please look at this.  If you want to discuss on gitter, please suggest a time between 9 AM and 9 PM California time.  There's not much time left before 8.9 if I'm going to make non trivial changes.

ejgallego commented:
@jfehrle , there is something I am not understanding. Why `idetop` has to know about the diff goals stuff?

That seems strange. I would suppose that the toplevels would just call the proper goal printing routine and it would the routine itself who would make the choice. Semantic information should be properly encapsulated.

Ok, I did a review, and I think all looks good but I am convinced there is a problem in `Proof_diffs`: the type of `match_goals` is incorrect, it should take `constr_expr` not `glob_expr`. This is indeed because the implementation of `to_constr` is incorrect. You should instead just use `Constrextern.extern_constr` and variants. Unfortunately this is a serious problem and it must be corrected before the code can go in. While you are at if please don't use exceptions for control in match_goals, but a boolean or even a richer type.

Please also remove the debug code from Proof_diffs. It should either go to `debug_printers`, or you can use serialization.

jfehrle commented:
> there is something I am not understanding. Why idetop has to know about the diff goals stuff?

At the least, I think this code should select the old proof step rather than having lower levels pick it.  And, IIRC, I got circular dependencies when I tried to call `Stm.get_prev_proof` from `Printer` or `Proof_diffs`.

I can try to move more of the diff-related code here into Proof_diffs.  But trying to hide all awareness of diffs here means scattering the changes widely and maybe replicating some routines.  I think the loss of locality would make it harder to follow what's happening.  I'd probably need to change `Proof.map_structured_proof`, `Idetop.process_goal`, then the print routines the latter calls: `pr_goal_concl_style_env` and `pr_compacted_decl`, and pass various wrapper variables hither and yon.  Trying to hide details is not always an improvement.

> the type of match_goals is incorrect, it should take constr_expr not glob_expr. This is indeed because the implementation of to_constr is incorrect. You should instead just use Constrextern.extern_constr and variants. Unfortunately this is a serious problem and it must be corrected before the code can go in.

Would you explain the difference between the two types and their uses, and why it is a serious problem?  Otherwise I've learned nothing.

This is a lot of stuff to change.

ejgallego commented:
Will answer later on idetop as it is not so urgent. Wrt glob_constr, we decided some time ago that it is an internal datatype. Even if it works for your uses you should not rely on it. A very famous plug-in was implement over that type and the pain was famous. I wrote an explanation some time ago but it had been lost in the mail archives. Hope to do a blog post some day 

Thus, you can either perform the matching directly on constr (which seems like the best choice for the purpose I think) or in constr_expr, which is the AST pre-printing. Note that the change should be straightforward, it looks harder than it actually is.

ejgallego commented:
On idetop: indeed the upper layer should select the previous state but then delegate to the printing routines what to print. Thus, I think there should not be any conditional on the proof diffs state.

ejgallego commented:
I don't see how this change would replicate code, as far as I can see it should have the contrary effect.

Regarding glob_constr, it is really a kind of low-level AST, needed by some particular implementation reasons , but indeed it supposes some limitations as it is generated without input from the pretyper. This impedes to use type-based resolution for many things. But the short story is, don't use that type. And please file a big from the part of the codebase you got that code as it must be fixed.

jfehrle commented:
I tried moving `export_pre_goals`, `process_goal_diffs`, etc. from `idetop.ml` into `proof_diffs.ml`.   First I get the error "Unbound module Interface".  Then, if I add, say, `let open Interface in` in the appropriate place, I still get the same error message, but now it's on the `let open`.  It seems like these 2 routines have to remain in idetop.ml because they rely on `Interface`.

Defining identical types `goal`, `pre_goals` and `goals` in Proof_diffs seems to make them incompatible with the same types in Interface.  My impression is that defining types identically in 2 different files makes them incompatible--the OCaml doc doesn't say anything about this.  That makes it mighty hard to use!

I don't fully understand how export_pre_goals works, in particular `Interface.whatever` and especially `pgs.Proof.whatever`.  So it becomes a long session of guesswork and experimentation, and even spending an hour or two researching it on the web is often unhelpful.

I don't understand why `Proof.map_structured_proof` refers to `zipper` rather than `stack`.  It's confusing.

So--can you explain in some detail how I can do what you're suggesting?  Or shall we let it be?

jfehrle commented:
Furthermore, `process_goal` has to remain in idetop.ml; it can't go into proof_diffs.ml because it calls routines in printer.ml.  For efficiency reasons, it's better to call `Proof_diffs.make_goal_map` once for each call to `goals` rather than creating it repeatedly for each call to `process_goal`.  That would mean keeping most of the diff code in idetop.ml.

Zimmi48 commented:
>My impression is that defining types identically in 2 different files makes them incompatible

That is correct. It is true in Coq too.

>Or shall we let it be?

Indeed, I think that @ejgallego views this as less of a concern than the question of `glob_constr` vs `constr` or `constr_expr`.

ejgallego commented:
I think you are trying to do a different, way more complicated thing to what I had in mind, moving interface to printer is not going to work obviously.

jfehrle commented:
I changed this to use `const_expr` instead of `glob_constr` as you requested.  I removed the debug code from `match_goals` (though I kept `db_goal_map`).  I didn't know what to do for idetop.

The exceptions raised in `match_goals` are for violations of the routine's assumptions, which I believe won't happen.  If they are raised, they should be shown to the user and reported as bugs.  I wouldn't change that.  (Also, FWIW, it seems to me that ending the tree traversal when I get one of these errors would mean checking after every call to `constr_expr`, so many lines would change for an improbable error.  I suppose I could set a `string ref` to indicate an error was found, but in that case it seems odd to continue the traversal anyway.)

Please take another look.  If it looks OK, please merge it.  And if not, let me know.

ejgallego commented:
> I changed this to use `const_expr` instead of `glob_constr` as you requested.

After thinking a bit more, wouldn't it make sense to do the matching over `Constr.t` directly? Think about it, you may want to refine it later, as matching may be more precise over the former.

I still have a gripe about the use of imperative code in `Proof_diffs`. That style is a bit far from the OCaml used in code, but what should we do? If I understand correctly you don't have a lot of OCaml experience, so it will take more iterations to get the code to a more idiomatic style. I was meaning to pass the map itself as a functional value; same for idetop, with the right higher-order pattern you can remove the conditional.

All in all, I don't oppose getting the code as-is, despite some OCaml annoyances in general the patch looks pretty good. This has already the green tick of two devs; but I am not sure they looked at the totality of the patch. I think I will let 2/3 days for @herbelin to look at the current version, and unless he complains I'll merge.

jfehrle commented:
OK, sounds good.

> After thinking a bit more, wouldn't it make sense to do the matching over Constr.t directly?

I don't know what the difference is between using Constr.t and const_expr, so I have no opinion.  Also, the matching is conceptually quite simple even though it's 150+ lines of code, not sure what might be gained.

> I was meaning to pass the map itself as a functional value; same for idetop, with the right higher-order pattern you can remove the conditional.

So something like add the map and the old proof as the initial arguments to `process_goal`, then curry those values for the diffs case?  I don't think I can add optional arguments to `process_goal` without changing the declaration for `Proof.map_structured_proof`, right?  (which I thought I should not change)

Thanks for the review.

ejgallego commented:
> I don't know what the difference is between using Constr.t and const_expr, so I have no opinion. Also, the matching is conceptually quite simple even though it's 150+ lines of code, not sure what might be gained.

There is a bit of text here: https://github.com/ejgallego/coq-serapi/blob/v8.8/FAQ.md#how-many-asts-does-coq-have also in the genarg.mli header: https://github.com/coq/coq/blob/master/lib/genarg.mli

_"hellish"_ could be a good way to explain it. Basically in constr you have less elaborated terms, so you gain a "purer" form to work with.

Regarding the changes, indeed, you could try like that; I don't think changing `Proof.map_structured_proof` is frowned upon. To what me regards I think that the API in proof is far from finish so of course it can take improvements from users.

Thanks for the patch and your patience.

ejgallego commented:
Last call for @herbelin 's comments [or request for more time], otherwise I shall merge tomorrow.
